File: [0019][Medium][Remove_Nth_Node_From_End_of_List][LinkedList_2_pointers]/Remove_Nth_Node_From_End_of_List_2.py
#Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# solution 3, recursion
class Solution:
    def removeNthFromEnd(self, head, n):
        global i 
        if head is None:
            i=0
            return None
        head.next = self.removeNthFromEnd(head.next,n)
        i+=1
        return head.next if i==n else head

File: [0019][Medium][Remove_Nth_Node_From_End_of_List][LinkedList_2_pointers]/Remove_Nth_Node_From_End_of_List_2.java

import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.LinkedList;

//Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

// solution 2, loop once
//consuming time: 0 ms,consuming memory: 37.9 MB,
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode first = dummy;
        ListNode second = dummy;
        // Advances first pointer so that the gap between first and second is n nodes apart
        for (int i = 1; i <= n + 1; i++) {
            first = first.next;
        }
        // Move first to the end, maintaining the gap
        while (first != null) {
            first = first.next;
            second = second.next;
        }
        second.next = second.next.next;
        return dummy.next;
    }
}